File: src/services/api/events.ts

import { API_BASE_URL, handleResponse } from './base';
import { Event, EventStaff } from '@/types';

// Get all events
export const getEvents = async (): Promise<Event[]> => {
  try {
    const response = await fetch(`${API_BASE_URL}/events`, {
      credentials: 'include',
    });

    if (!response.ok) {
      throw new Error(`Error fetching events: ${response.status}`);
    }

    const data = await response.json();
    
    if (!Array.isArray(data)) {
      console.error("API did not return an array for events", data);
      return [];
    }
    
    return data.map((event: any) => ({
      id: event.EventID,
      name: event.EventName,
      date: event.EventDate,
      location: event.Location,
      description: event.Description,
      clientName: event.ClientName || "",
      clientContact: event.ClientContact || "",
      status: event.Status || "Upcoming",
      fee: event.Fee || 0,
      eventType: event.EventType || "",
    }));
  } catch (error) {
    console.error("Error in getEvents:", error);
    return [];
  }
};

// Create a new event
export const createEvent = async (eventData: EventFormData): Promise<Event> => {
  const response = await fetch(`${API_BASE_URL}/events`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      event_name: eventData.name,
      event_date: eventData.date,
      location: eventData.location,
      description: eventData.description,
      client_name: eventData.clientName,
      client_contact: eventData.clientContact,
      status: eventData.status,
      fee: eventData.fee,
      event_type: eventData.eventType,
    }),
    credentials: 'include',
  });
  return handleResponse(response);
};

// Update an existing event
export const updateEvent = async (eventId: number, eventData: EventFormData): Promise<Event> => {
  const response = await fetch(`${API_BASE_URL}/events/${eventId}`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      event_name: eventData.name,
      event_date: eventData.date,
      location: eventData.location,
      description: eventData.description,
      client_name: eventData.clientName,
      client_contact: eventData.clientContact,
      status: eventData.status,
      fee: eventData.fee,
      event_type: eventData.eventType,
    }),
    credentials: 'include',
  });
  return handleResponse(response);
};

// Delete an event
export const deleteEvent = async (eventId: number): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/events/${eventId}`, {
    method: 'DELETE',
    credentials: 'include',
  });
  return handleResponse(response);
};

// Get staff assigned to an event
export const getEventStaff = async (eventId: number): Promise<EventStaff[]> => {
  try {
    const response = await fetch(`${API_BASE_URL}/events/${eventId}/staff`, {
      credentials: 'include',
    });

    if (!response.ok) {
      throw new Error(`Error fetching event staff: ${response.status}`);
    }

    const data = await response.json();
    
    if (!Array.isArray(data)) {
      console.error("API did not return an array for event staff", data);
      return [];
    }
    
    return data.map((staff: any) => ({
      id: staff.EventStaffID,
      eventId: staff.EventID,
      trainerId: staff.TrainerID,
      trainerName: staff.TrainerName || "",
      role: staff.Role || "",
      confirmed: staff.Confirmed || false,
    }));
  } catch (error) {
    console.error("Error in getEventStaff:", error);
    return [];
  }
};

// Assign staff to an event
export const assignStaffToEvent = async (eventId: number, trainerId: number, role: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/events/${eventId}/staff`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      trainer_id: trainerId,
      role: role,
    }),
    credentials: 'include',
  });
  return handleResponse(response);
};

// Remove staff from an event
export const removeStaffFromEvent = async (eventId: number, staffId: number): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/events/${eventId}/staff/${staffId}`, {
    method: 'DELETE',
    credentials: 'include',
  });
  return handleResponse(response);
};

File: src/api/routes/events.py
from flask import Blueprint, request, jsonify
from utils.db import get_db_connection

events_bp = Blueprint('events', __name__)

# Get all events
@events_bp.route('/', methods=['GET'])
def get_events():
    conn = get_db_connection()
    if not conn:
        return jsonify({'error': 'Database connection failed'}), 500

    try:
        cursor = conn.cursor(dictionary=True)
        query = "SELECT * FROM Events"
        cursor.execute(query)
        events = cursor.fetchall()
        cursor.close()
        conn.close()
        return jsonify(events)
    except Exception as e:
        print(f"Error fetching events: {e}")
        if conn:
            conn.close()
        return jsonify({'error': str(e)}), 500

# Create a new event
@events_bp.route('', methods=['POST'])
def create_event():
    data = request.json
    conn = get_db_connection()
    if not conn:
        return jsonify({'error': 'Database connection failed'}), 500

    try:
        cursor = conn.cursor()
        query = """
        INSERT INTO Events (EventName, EventDate, Location, Description, ClientName, ClientContact, Status, Fee, EventType)
        VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
        """
        values = (
            data.get('eventname'),
            data.get('eventdate'),
            data.get('location'),
            data.get('description'),
            data.get('clientName'),
            data.get('clientContact'),
            data.get('status'),
            data.get('fee'),
            data.get('eventType')
        )
        cursor.execute(query, values)
        conn.commit()
        cursor.close()
        conn.close()
        return jsonify({'success': True, 'message': 'Event created successfully'})
    except Exception as e:
        print(f"Error creating event: {e}")
        if conn:
            conn.close()
        return jsonify({'error': str(e)}), 500

# Update an existing event
@events_bp.route('/<int:event_id>', methods=['PUT'])
def update_event(event_id):
    data = request.json
    conn = get_db_connection()
    if not conn:
        return jsonify({'error': 'Database connection failed'}), 500

    try:
        cursor = conn.cursor()
        query = """
        UPDATE Events
        SET EventName = %s, EventDate = %s, Location = %s, Description = %s, ClientName = %s, ClientContact = %s, Status = %s, Fee = %s, EventType = %s
        WHERE EventID = %s
        """
        values = (
            data.get('name'),
            data.get('date'),
            data.get('location'),
            data.get('description'),
            data.get('clientName'),
            data.get('clientContact'),
            data.get('status'),
            data.get('fee'),
            data.get('eventType'),
            event_id
        )
        cursor.execute(query, values)
        conn.commit()
        if cursor.rowcount == 0:
            conn.close()
            return jsonify({'error': 'Event not found'}), 404
        cursor.close()
        conn.close()
        return jsonify({'success': True, 'message': 'Event updated successfully'})
    except Exception as e:
        print(f"Error updating event: {e}")
        if conn:
            conn.close()
        return jsonify({'error': str(e)}), 500

# Delete an event
@events_bp.route('/<int:event_id>', methods=['DELETE'])
def delete_event(event_id):
    conn = get_db_connection()
    if not conn:
        return jsonify({'error': 'Database connection failed'}), 500

    try:
        cursor = conn.cursor()
        query = "DELETE FROM Events WHERE EventID = %s"
        cursor.execute(query, (event_id,))
        conn.commit()
        if cursor.rowcount == 0:
            conn.close()
            return jsonify({'error': 'Event not found'}), 404
        cursor.close()
        conn.close()
        return jsonify({'success': True, 'message': 'Event deleted successfully'})
    except Exception as e:
        print(f"Error deleting event: {e}")
        if conn:
            conn.close()
        return jsonify({'error': str(e)}), 500

# Get staff assigned to an event
@events_bp.route('/<int:event_id>/staff', methods=['GET'])
def get_event_staff(event_id):
    conn = get_db_connection()
    if not conn:
        return jsonify({'error': 'Database connection failed'}), 500

    try:
        cursor = conn.cursor(dictionary=True)
        query = "SELECT * FROM EventStaff WHERE EventID = %s"
        cursor.execute(query, (event_id,))
        staff = cursor.fetchall()
        cursor.close()
        conn.close()
        return jsonify(staff)
    except Exception as e:
        print(f"Error fetching event staff: {e}")
        if conn:
            conn.close()
        return jsonify({'error': str(e)}), 500

# Assign staff to an event
@events_bp.route('/<int:event_id>/staff', methods=['POST'])
def assign_staff_to_event(event_id):
    data = request.json
    conn = get_db_connection()
    if not conn:
        return jsonify({'error': 'Database connection failed'}), 500

    try:
        cursor = conn.cursor()
        query = """
        INSERT INTO EventStaff (EventID, TrainerID, Role)
        VALUES (%s, %s, %s)
        """
        values = (event_id, data.get('trainerId'), data.get('role'))
        cursor.execute(query, values)
        conn.commit()
        cursor.close()
        conn.close()
        return jsonify({'success': True, 'message': 'Staff assigned to event successfully'})
    except Exception as e:
        print(f"Error assigning staff to event: {e}")
        if conn:
            conn.close()
        return jsonify({'error': str(e)}), 500

# Remove staff from an event
@events_bp.route('/<int:event_id>/staff/<int:staff_id>', methods=['DELETE'])
def remove_staff_from_event(event_id, staff_id):
    conn = get_db_connection()
    if not conn:
        return jsonify({'error': 'Database connection failed'}), 500

    try:
        cursor = conn.cursor()
        query = "DELETE FROM EventStaff WHERE EventID = %s AND EventStaffID = %s"
        cursor.execute(query, (event_id, staff_id))
        conn.commit()
        if cursor.rowcount == 0:
            conn.close()
            return jsonify({'error': 'Staff not found for this event'}), 404
        cursor.close()
        conn.close()
        return jsonify({'success': True, 'message': 'Staff removed from event successfully'})
    except Exception as e:
        print(f"Error removing staff from event: {e}")
        if conn:
            conn.close()
        return jsonify({'error': str(e)}), 500